src/util/ObservableValue.hpp

```diff
@@ -0,0 +1,520 @@
+     */
+    ObservableValue()
+        requires std::default_initializable<T>
+        : value_{}
+    {
+    }
+
+    ObservableValue(ObservableValue const&) = delete;
+    ObservableValue(ObservableValue&&) = default;
+    ObservableValue&
+    operator=(ObservableValue const&) = delete;
+    ObservableValue&
+    operator=(ObservableValue&&) = default;
+
+    /**
+     * @brief Assignment operator that updates atomic value and notifies observers.
+     *
+     * Uses atomic compare-and-swap to update the value and notifies observers
+     * only if the value actually changed.
+     *
+     * @param val New value
+     * @return Reference to this object for chaining
+     */
+    ObservableValue&
+    operator=(std::convertible_to<T> auto&& val)
+    {
+        set(std::forward<decltype(val)>(val));
+        return *this;
+    }
+
+    /**
+     * @brief Provides deferred notification access to the atomic value.
+     *
+     * Returns an AtomicObservableGuard that allows modification of the atomic
+     * with notification deferred until the guard is destroyed.
+     *
+     * @return AtomicObservableGuard for deferred notification
+     */
+    [[nodiscard]] AtomicObservableGuard
+    operator->()
+    {
+        return {*this};
+    }
+
+    /**
+     * @brief Gets the current atomic value.
+     * @return Current value stored in the atomic
+     */
+    [[nodiscard]] T
+    get() const
+    {
+        return value_.load();
+    }
+
+    /**
+     * @brief Implicit conversion to the current atomic value.
+     * @return Current value stored in the atomic
+     */
+    [[nodiscard]]
+    operator T() const
+    {
+        return get();
+    }
+
+    /**
+     * @brief Sets a new atomic value and notifies observers if changed.
+     *
+     * Uses atomic compare-and-swap to update the value. Notifies all observers
+     * if the value actually changed.
+     *
+     * @param val New value
+     */
+    void
+    set(std::convertible_to<T> auto&& val)
+    {
+        T newValue = std::forward<decltype(val)>(val);
+        T oldValue = value_.load();
+
+        // Use compare-and-swap to atomically update
+        while (!value_.compare_exchange_weak(oldValue, newValue)) {
+            // compare_exchange_weak updates oldValue with current value on failure
+            // Continue until we succeed
+        }
+
+        // Notify observers if we actually changed the value
+        // Note: oldValue now contains the actual previous value that was replaced
+        if (oldValue != newValue) {
+            this->notifyObservers(newValue);
+        }
+    }
+
+    /**
+     * @brief Provides direct access to the underlying atomic.
+     *
+     * Use with caution - direct atomic operations bypass observation.
+     * Prefer using set() or the guard mechanism for observable changes.
+     *
+     * @return Reference to the underlying atomic
+     */
+    [[nodiscard]] std::atomic<T>&
```

Comment:
This should probably just be removed unless it is really used and needed somewhere.